---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---
```{r}
knitr::opts_knit$set(root.dir = "path/to/working/directionary")
knitr::opts_chunk$set(warning = FALSE,
                      error = TRUE,
                      echo = FALSE)
```
```{r , eval=FALSE, include=FALSE}
source("utilities.R")
library(rstudioapi)
knit_with_date(rstudioapi::getActiveDocumentContext()$path)
```

```{r}
library(dplyr)
library(stringr)
library(data.table)
library(readr)
library(tidyverse)
library(readxl)
library(gggenes)
# Setting up enviroment for work folder
library(here)
d <- function(x, raw = TRUE){paste0(here(),"/data/",ifelse(raw, "raw/", "processed/"), x)}
o <- function(x){paste(here(), x, sep = "/output/")}
f <- function(x){paste(here(), x, sep = "/figures/")}
s <- function(x){paste(here(), x, sep = "/scripts/")}

```

## Goal: Do the operon identification data.table notation
## Date: 23 maj 2021
## Author: Soren H


```{r}
filename_psiblast <- "cellulose1"
perc_id = 20
max_dist_prok = 8
max_dist_gene = 2000
min_genes = 2
```



```{r}
magstats <- fread(d("magstats.tsv"), sep = "\t")

query_metadata <- excel_sheets(d("/Query_figur.xlsx")) %>% 
  sapply(function(X) read_xlsx(d("/Query_figur.xlsx"), sheet = X, skip = 1), USE.NAMES = T) %>% 
  lapply(as.data.frame) %>% 
  `[`(!(names(.) %in% c("Abbreveations", "HA_S_pyogenes"))) %>% 
  rbindlist()

gff <- fread(d("gff.tsv")) %>% 
  mutate(Target_label = paste(ID, ProkkaNO, sep="_"))
```


```{r}
psiblast <- fread(
  file = d(paste0("/psiblast/", filename_psiblast, ".txt")), 
  header = FALSE, fill=TRUE, sep = "\t", 
  col.names = c("Query_label", "Target_label","Percent_identity","align_length","mismatch", "gap_opens","start_query", "end_query", "start_target", "end_target","E_value", "Bit score")
) 
psiblast <- psiblast %>% 
  subset(!is.na(E_value)) %>%   
  separate(Target_label, c("ID", "ProkkaNO"), sep = -5, remove = FALSE, extra = "merge") %>%
  mutate(ID = str_sub(ID, start = 1, end = -2)) %>%
  filter(Percent_identity > perc_id) %>% 
  group_by(Target_label) %>% 
  filter(`Bit score` == max(`Bit score`)) %>% 
  filter(!duplicated(Target_label)) %>%   
  left_join(gff, by = c("Target_label", "ProkkaNO", "ID"), keep = FALSE) %>%
  left_join(magstats, by = "ID", keep = FALSE) %>% 
  mutate(ProkkaINT = as.numeric(ProkkaNO)) %>% 
  as.data.table() %>% 
  arrange(Target_label, ProkkaINT) %>% 
  left_join(x=., y=query_metadata[, c("Genename", "Function")], by = c("Query_label" = "Genename"))
#rm(gff, magstats)
```



```{r}

# [
#   Calculate distance to anterior and posterior hit for both prokka and nucleotides
# ][
#   Remove all hits with no prokka/gene distance below threshhold
# ][
#   Search for operon start genes (Only prokka/gene distance to preceding hit is within threshold)
# ]
psiblast <- psiblast[, `:=`(
                   ante_prok = replace_na(ProkkaINT - shift(ProkkaINT, 1), 10000),
                   post_prok = replace_na(shift(ProkkaINT, -1) - ProkkaINT, 10000),
                   ante_gene = replace_na(start - shift(end, 1), 10000),
                   post_gene = replace_na(shift(start, -1) - end, 10000)
                 )
                 , by=.(seqname, ID)
                 ][
                   {ante_prok < max_dist_prok | post_prok < max_dist_prok} & {ante_gene < max_dist_gene | post_gene < max_dist_gene}
                 ][,
                   operon_place := case_when(
                     (post_prok < max_dist_prok & post_gene < max_dist_gene) & 
                     !(ante_prok < max_dist_prok & ante_gene < max_dist_gene) ~ "start",
                     TRUE ~ ".")
                 ]

# Assigning unique number to each operon
psiblast[, "operon"] <- 0
psiblast[, "operon"] <- as.numeric(psiblast$operon)
for (i  in 1:nrow(psiblast)) {
  if (psiblast[i,]$operon_place == "start"){
    psiblast[i, "operon"] <- max(psiblast[, "operon"], na.rm = TRUE) + 1
  }
  psiblast[i,"operon"] <- max(psiblast[, "operon"], na.rm = TRUE)
}
psiblast <- psiblast %>% 
  group_by(operon) %>% 
  filter(length(unique(Query_label)) >= min_genes) %>% 
  mutate(
    end = end - min(start),
    start = start - min(start)
    ) %>% 
  mutate(
    strand = case_when(
      strand=="+" ~ 1,
      strand=="-" ~ -1,
      TRUE ~ as.numeric(strand)
      )
  )
```

Plotting using GGgenes

```{r}
pdf(f("/gggenes_test_cellulose.pdf"), height=25, width = 15)
ggplot(psiblast, aes(xmin=start, xmax=end, y = ID, fill=Function, forward = strand, label = Query_label)) + 
  geom_gene_arrow(arrowhead_height = unit(4, "mm"), arrowhead_width = unit(2, "mm"), arrow_body_height = unit(4, "mm")) + 
  geom_gene_label(align = "left", padding.x = unit(2, "mm")) +
  facet_wrap(~ operon,scales = "free", ncol=1) + 
  scale_fill_brewer(palette = "Set3") +
  theme_genes()
dev.off()
```






















